### DISCUSSION TITLE
TCP Plugin for RFC 9329 and draft-smyslov-ipsecme-ikev2-reliable-transport-01

### DISCUSSION THREAD

**Discussion by ghost** [strongswan/strongswan#2210] [General]
I'm attempting to implement a rough version of RFC 9329 and draft-smyslov-ipsecme-ikev2-reliable-transport-01 by adding a TCP plugin to libcharon. The rough idea was that the plugin would observe the IKE_INIT messages for the correct notify or bytes and would route messages to UDP or TCP based on the configuration; either full TCP encap or IKE only TCP encap. I've been following the socket_default as a rough guide to build the TCP socket on but upon build and installation, Charon throws this error:

00[LIB] plugin 'tcpd' failed to load: /usr/lib/ipsec/plugins/libstrongswan-tcpd.so: undefined symbol: tcp_decouple_socket_create

Here is the associated header file:

#ifndef TCP_DECOUPLE_SOCKET_H_
#define TCP_DECOUPLE_SOCKET_H_
typedef struct tcp_decouple_socket tcp_decouple_socket;
#include <network/socket.h>
/**
 * A socket implementation binding to TCP port 4500 for IKE Messages.
 */
struct tcp_decouple_socket {
    /**
     * Implements the socket_t interface.
     */
    socket_t socket;
};
/**
 * Create a tcp_decouple_socket instance.
 */
tcp_decouple_socket *tcp_decouple_socket_create();
#endif /** TCP_DECOUPLE_SOCKET_H_ @}*/

My question being, is this a good approach? And why might Charon be throwing this error?

**Comment by tobiasbrunner** [comment]
Not sure about the approach, but the error is clearly something you caused yourself somehow.
